File: backend/timeline_time_parser.py
import re
import csv
import io
from typing import Optional, List, Dict, Any, Tuple

def parse_time_to_seconds(time_str: str, allow_relative: bool = False) -> Optional[float]:
    if not time_str:
        return None
        
    t = str(time_str).strip().lower()
    if not t:
        return None
        
    is_relative = t.startswith('+')
    str_to_parse = t[1:] if is_relative else t
    
    if is_relative and not allow_relative:
        return None
        
    # handle raw numbers or simple decimals
    if re.fullmatch(r'^\d+(\.\d+)?$', str_to_parse):
        return float(str_to_parse)
        
    # Check if Excel exported an AM/PM clock time
    if 'am' in t or 'pm' in t:
        return None

    # hh:mm:ss or mm:ss or mm:ss.ms
    hms_match = re.fullmatch(r'^(?:(?:(\d+):)?(\d{1,2}):)?(\d{1,2}(?:\.\d+)?)$', str_to_parse)
    if hms_match:
        p1 = hms_match.group(1)
        p2 = hms_match.group(2)
        p3 = hms_match.group(3)
        
        if p1 is not None:
            # 3 parts: HH:MM:SS
            h = int(p1)
            m = int(p2)
            s = float(p3)
            return h * 3600 + m * 60 + s
        elif p2 is not None:
            # 2 parts: MM:SS
            m = int(p2)
            s = float(p3)
            return m * 60 + s
        else:
            # 1 part: SS
            return float(p3)
            
    # Handle s, m, min, sec formats
    min_sec_match = re.fullmatch(r'^(?:(\d+)\s*(?:m|min|mins|minutes))?\s*(?:(\d+(?:\.\d+)?)\s*(?:s|sec|secs|seconds))?$', str_to_parse)
    if min_sec_match and (min_sec_match.group(1) or min_sec_match.group(2)):
        m = int(min_sec_match.group(1) or "0")
        s = float(min_sec_match.group(2) or "0")
        return m * 60 + s
        
    return None

def parse_timeline_csv(csv_text: str, timeline_type: str) -> Tuple[bool, List[Dict[str, Any]], float, List[str], List[str], str]:
    """
    timeline_type: 'image', 'media', 'video'
    Returns: (success, rows, total_duration, errors, warnings, normalized_csv)
    """
    lines = [l for l in csv_text.splitlines() if l.strip()]
    if len(lines) < 2:
        return False, [], 0.0, ["Timeline CSV must contain a header and at least one row."], [], ""
        
    reader = csv.reader(io.StringIO(lines[0]))
    headers = []
    try:
        headers = next(reader)
        headers = [h.strip().lower() for h in headers]
    except StopIteration:
        pass
        
    start_col, end_col, file_col, text_col = -1, -1, -1, -1
    start_aliases = ['start', 'start_time', 'starttime']
    end_aliases = ['end', 'end_time', 'endtime']
    file_aliases = ['image', 'asset', 'video', 'file', 'media', 'path']
    text_aliases = ['text', 'caption', 'script']
    
    for i, h in enumerate(headers):
        if h in start_aliases: start_col = i
        elif h in end_aliases: end_col = i
        elif h in file_aliases: file_col = i
        elif h in text_aliases: text_col = i
        
    if start_col == -1 or end_col == -1 or file_col == -1:
        return False, [], 0.0, ["CSV missing required columns. Ensure it has start, end, and file/asset/video/image columns."], [], ""
        
    parsed_rows = []
    errors = []
    warnings = []
    previous_row_end = 0.0
    
    reader = csv.reader(io.StringIO("\n".join(lines[1:])))
    for idx, columns in enumerate(reader):
        row_num = idx + 2
        
        if len(columns) == 1 and columns[0].strip() == "":
            continue
            
        start_str = columns[start_col].strip() if start_col < len(columns) else ""
        end_str = columns[end_col].strip() if end_col < len(columns) else ""
        file_val = columns[file_col].strip() if file_col < len(columns) else ""
        text_val = columns[text_col].strip() if text_col != -1 and text_col < len(columns) else ""
        
        if not file_val:
            errors.append(f"Row {row_num}: File is missing.")
            continue
            
        if not end_str:
            errors.append(f"Row {row_num}: Missing end time.")
            continue
            
        if 'am' in end_str.lower() or 'pm' in end_str.lower() or 'am' in start_str.lower() or 'pm' in start_str.lower():
            errors.append("Excel converted your timestamp into a clock time. Format the start/end columns as Text before exporting CSV.")
            continue
            
        is_relative_end = end_str.startswith('+')
        parsed_end = parse_time_to_seconds(end_str, allow_relative=True)
        
        if parsed_end is None:
            errors.append(f"Row {row_num}: Invalid time format \"{end_str}\" in end column.")
            continue
            
        if not start_str:
            if len(parsed_rows) == 0:
                start_sec = 0.0
            else:
                start_sec = previous_row_end
        else:
            parsed_start = parse_time_to_seconds(start_str, allow_relative=False)
            if parsed_start is None:
                errors.append(f"Row {row_num}: Invalid time format \"{start_str}\" in start column.")
                continue
            start_sec = parsed_start
            
        if is_relative_end:
            end_sec = start_sec + parsed_end
        else:
            end_sec = parsed_end
            
        if end_sec <= start_sec:
            errors.append(f"Row {row_num}: End time must be greater than start time.")
            continue
            
        if len(parsed_rows) > 0:
            if start_sec < previous_row_end:
                errors.append(f"Row {row_num} overlaps the previous row.")
                continue
            if start_sec > previous_row_end:
                warnings.append(f"Row {row_num} starts after the previous row ends. This creates a gap.")
            
        parsed_rows.append({
            "start": start_sec,
            "end": end_sec,
            "file": file_val,
            "text": text_val
        })
        previous_row_end = end_sec
        
    if errors:
        return False, [], 0.0, errors, warnings, ""
        
    def escape_csv_field(field: str) -> str:
        if not field: return ""
        f = str(field)
        if "," in f or '"' in f or "\n" in f or "\r" in f:
            return '"{}"'.format(f.replace('"', '""'))
        return f

    out_csv_lines = []
    if timeline_type == 'image':
        out_csv_lines.append("image,start,end,text")
        for r in parsed_rows:
            out_csv_lines.append(f"{escape_csv_field(r['file'])},{r['start']},{r['end']},{escape_csv_field(r['text'])}")
    elif timeline_type == 'media':
        out_csv_lines.append("start,end,asset,text")
        for r in parsed_rows:
            out_csv_lines.append(f"{r['start']},{r['end']},{escape_csv_field(r['file'])},{escape_csv_field(r['text'])}")
    else:
        out_csv_lines.append("start,end,video")
        for r in parsed_rows:
            out_csv_lines.append(f"{r['start']},{r['end']},{escape_csv_field(r['file'])}")
            
    total_duration = parsed_rows[-1]["end"] if parsed_rows else 0.0
    normalized_csv = "\n".join(out_csv_lines)
    
    return True, parsed_rows, total_duration, [], warnings, normalized_csv

File: frontend/src/utils/timelineTimeParser.ts
export type ParsedRow = {
  start: number;
  end: number;
  file: string;
  text: string;
};

export type ParseResult = {
  success: boolean;
  rows: ParsedRow[];
  totalDurationSeconds: number; // raw total duration
  errors: string[];
  warnings: string[];
  normalizedCsv: string; // The normalized CSV string that can be uploaded
};

export function parseTimeToSeconds(timeStr: string, options?: { allowRelative?: boolean }): number | null {
  if (!timeStr) return null;
  const t = timeStr.trim().toLowerCase();
  
  const isRelative = t.startsWith('+');
  const strToParse = isRelative ? t.substring(1) : t;
  
  if (isRelative && !options?.allowRelative) {
    return null; // rejected
  }

  // handle raw numbers or simple decimals
  if (/^\d+(\.\d+)?$/.test(strToParse)) {
    return parseFloat(strToParse);
  }

  // Check if Excel exported an AM/PM clock time
  if (t.includes('am') || t.includes('pm')) {
    return NaN; // Let validation handle this specifically later, or throw
  }

  // hh:mm:ss or mm:ss or mm:ss.ms
  const hmsRegex = /^(?:(?:(\d+):)?(\d{1,2}):)?(\d{1,2}(?:\.\d+)?)$/;
  const hmsMatch = strToParse.match(hmsRegex);
  if (hmsMatch) {
    const p1 = hmsMatch[1];
    const p2 = hmsMatch[2];
    const p3 = hmsMatch[3];

    // If 3 parts: HH:MM:SS
    if (p1 !== undefined) {
      const h = parseInt(p1, 10);
      const m = parseInt(p2, 10);
      const s = parseFloat(p3);
      return h * 3600 + m * 60 + s;
    }
    // If 2 parts: MM:SS
    else if (p2 !== undefined) {
      const m = parseInt(p2, 10);
      const s = parseFloat(p3);
      return m * 60 + s;
    }
    // If 1 part: SS (caught by simple decimals above, but just in case)
    else {
      return parseFloat(p3);
    }
  }

  // Handle s, m, min, sec formats
  const minSecRegex = /^(?:(\d+)\s*(?:m|min|mins|minutes))?\s*(?:(\d+(?:\.\d+)?)\s*(?:s|sec|secs|seconds))?$/;
  const minSecMatch = strToParse.match(minSecRegex);
  if (minSecMatch && (minSecMatch[1] || minSecMatch[2])) {
    const m = parseInt(minSecMatch[1] || "0", 10);
    const s = parseFloat(minSecMatch[2] || "0");
    return m * 60 + s;
  }

  return null;
}

export function formatSecondsForDisplay(seconds: number): string {
  if (isNaN(seconds) || seconds < 0) return "0:00";
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  const s = seconds % 60;
  
  const sStr = s % 1 === 0 ? s.toString().padStart(2, '0') : s.toFixed(3).padStart(6, '0');
  
  if (h > 0) {
    return `${h}:${m.toString().padStart(2, '0')}:${sStr}`;
  }
  return `${m}:${sStr}`;
}

export function validateTimelineRows(rows: ParsedRow[]): string[] {
  const errors: string[] = [];
  rows.forEach((r, i) => {
    if (r.end <= r.start) {
      errors.push(`Row ${i + 2}: End time must be greater than start time.`);
    }
  });
  return errors;
}

// simple CSV parser handling quotes
export function parseCSVLine(line: string): string[] {
  const result = [];
  let current = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    if (char === '"') {
      if (inQuotes && line[i + 1] === '"') {
        current += '"';
        i++;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (char === ',' && !inQuotes) {
      result.push(current);
      current = "";
    } else {
      current += char;
    }
  }
  result.push(current);
  return result;
}

export function escapeCSVField(field: string): string {
  if (field === undefined || field === null) return "";
  const str = String(field);
  if (str.includes(',') || str.includes('"') || str.includes('\n') || str.includes('\r')) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

export function parseTimelineCsv(csvText: string, timelineType: 'image' | 'media' | 'video'): ParseResult {
  const lines = csvText.split(/\r?\n/).filter(l => l.trim() !== "");
  if (lines.length < 2) {
    return { success: false, rows: [], totalDurationSeconds: 0, errors: ["Timeline CSV must contain a header and at least one row."], warnings: [], normalizedCsv: "" };
  }

  const headers = parseCSVLine(lines[0]).map(h => h.trim().toLowerCase());
  
  // Find column indices
  let startCol = -1, endCol = -1, fileCol = -1, textCol = -1;
  const startAliases = ['start', 'start_time', 'starttime'];
  const endAliases = ['end', 'end_time', 'endtime'];
  const fileAliases = ['image', 'asset', 'video', 'file', 'media', 'path'];
  const textAliases = ['text', 'caption', 'script'];

  headers.forEach((h, i) => {
    if (startAliases.includes(h)) startCol = i;
    else if (endAliases.includes(h)) endCol = i;
    else if (fileAliases.includes(h)) fileCol = i;
    else if (textAliases.includes(h)) textCol = i;
  });

  if (startCol === -1 || endCol === -1 || fileCol === -1) {
    return { success: false, rows: [], totalDurationSeconds: 0, errors: ["CSV missing required columns. Ensure it has start, end, and file/asset/video/image columns."], warnings: [], normalizedCsv: "" };
  }

  const parsedRows: ParsedRow[] = [];
  const errors: string[] = [];
  const warnings: string[] = [];
  let previousRowEnd = 0;

  for (let i = 1; i < lines.length; i++) {
    const rowLine = lines[i];
    const columns = parseCSVLine(rowLine);
    const rowNum = i + 1;

    if (columns.length === 1 && columns[0].trim() === "") continue;

    const startStr = (columns[startCol] || "").trim();
    const endStr = (columns[endCol] || "").trim();
    const file = (columns[fileCol] || "").trim();
    const text = textCol !== -1 ? (columns[textCol] || "").trim() : "";

    if (!file && !text) {
      errors.push(`Row ${rowNum}: Both File and Text are missing. At least one is required.`);
      continue;
    }

    if (!endStr) {
      errors.push(`Row ${rowNum}: Missing end time.`);
      continue;
    }
    
    // Check for Excel AM/PM format
    if (endStr.toLowerCase().includes('am') || endStr.toLowerCase().includes('pm') ||
        startStr.toLowerCase().includes('am') || startStr.toLowerCase().includes('pm')) {
      errors.push(`Excel converted your timestamp into a clock time. Format the start/end columns as Text before exporting CSV.`);
      continue;
    }

    let startSec: number;
    let endSec: number;

    const isRelativeEnd = endStr.startsWith('+');
    const parsedEnd = parseTimeToSeconds(endStr, { allowRelative: true });

    if (parsedEnd === null || isNaN(parsedEnd)) {
      errors.push(`Row ${rowNum}: Invalid time format "${endStr}" in end column.`);
      continue;
    }

    if (!startStr) {
      if (parsedRows.length === 0) { // First valid row
        startSec = 0;
      } else {
        startSec = previousRowEnd;
      }
    } else {
      const parsedStart = parseTimeToSeconds(startStr, { allowRelative: false });
      if (parsedStart === null || isNaN(parsedStart)) {
        errors.push(`Row ${rowNum}: Invalid time format "${startStr}" in start column.`);
        continue;
      }
      startSec = parsedStart;
    }

    if (isRelativeEnd) {
      endSec = startSec + parsedEnd;
    } else {
      endSec = parsedEnd;
    }

    if (endSec <= startSec) {
      errors.push(`Row ${rowNum}: End time must be greater than start time.`);
      continue;
    }
    
    if (parsedRows.length > 0) {
      if (startSec < previousRowEnd) {
        errors.push(`Row ${rowNum} overlaps the previous row.`);
        continue;
      }
      if (startSec > previousRowEnd) {
        warnings.push(`Row ${rowNum} starts after the previous row ends. This creates a gap.`);
      }
    }

    parsedRows.push({ start: startSec, end: endSec, file, text });
    previousRowEnd = endSec;
  }

  if (errors.length > 0) {
    return { success: false, rows: [], totalDurationSeconds: 0, errors, warnings, normalizedCsv: "" };
  }
  
  if (warnings.length > 0) {
    // We can surface warnings through errors array if UI expects strings, 
    // but user says "allow it, but show warning". Let's format it so UI can show it but not block.
    // Actually the return signature doesn't have `warnings`. We can append to a global state or add to ParseResult.
    // Let's add warnings to the return result.
  }

  // Generate normalized CSV
  let outCsvLines = [];
  if (timelineType === 'image') {
    outCsvLines.push("image,start,end,text");
    parsedRows.forEach(r => outCsvLines.push(`${escapeCSVField(r.file)},${r.start},${r.end},${escapeCSVField(r.text)}`));
  } else if (timelineType === 'media') {
    outCsvLines.push("start,end,asset,text");
    parsedRows.forEach(r => outCsvLines.push(`${r.start},${r.end},${escapeCSVField(r.file)},${escapeCSVField(r.text)}`));
  } else {
    outCsvLines.push("start,end,video");
    parsedRows.forEach(r => outCsvLines.push(`${r.start},${r.end},${escapeCSVField(r.file)}`));
  }

  const totalDuration = parsedRows.length > 0 ? parsedRows[parsedRows.length - 1].end : 0;

  return {
    success: true,
    rows: parsedRows,
    totalDurationSeconds: totalDuration,
    errors: [],
    warnings: warnings,
    normalizedCsv: outCsvLines.join("\n")
  };
}
